From 5014e085a84b59dcea58624b9a9d7cc572e1fe63 Mon Sep 17 00:00:00 2001
From: Carl George <carl@george.computer>
Date: Tue, 19 Apr 2022 23:21:27 -0500
Subject: [PATCH 3/3] Add support for Python 3.6 through 3.11

I noticed this package in Fedora was failing to rebuild with Python
3.11.  `inspect.getargspec` has been removed.  `inspect.getfullargspec`
is the updated API, but returns a tuple with more properties.  With that
handled in a try/except block, the tests pass on 3.6.15, 3.7.13, 3.8.13,
3.9.12, 3.10.4, and 3.11.0a7.

(cherry picked from commit 503f1243999bf3c0d500f3e57639519733ae93a8)
---
 easyargs/main.py    | 5 ++++-
 easyargs/parsers.py | 6 +++++-
 setup.py            | 6 ++++++
 3 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/easyargs/main.py b/easyargs/main.py
index aea30d3..385959c 100755
--- a/easyargs/main.py
+++ b/easyargs/main.py
@@ -36,7 +36,10 @@ def create_sub_parser(parser, method_info):
 
     # Get the arguments for the method
     # TODO: do something sensible with args and varargs
-    args, _, _, defaults = inspect.getargspec(method)
+    try:
+        args, _, _, defaults, _, _, _ = inspect.getfullargspec(method)
+    except AttributeError:
+        args, _, _, defaults = inspect.getargspec(method)
 
     num_positional = len(args)
     if defaults is not None:
diff --git a/easyargs/parsers.py b/easyargs/parsers.py
index 8c9adea..e330e48 100644
--- a/easyargs/parsers.py
+++ b/easyargs/parsers.py
@@ -106,7 +106,11 @@ def function_parser(function, parser):
     main_text, params_help = parser_help_text(help_text)
 
     # Get the function information
-    args, varargs, keywords, defaults = inspect.getargspec(function)
+    try:
+        args, varargs, keywords, defaults, _, _, _ = inspect.getfullargspec(function)
+    except AttributeError:
+        args, varargs, keywords, defaults = inspect.getargspec(function)
+
     if args is None:
         args = []
 
diff --git a/setup.py b/setup.py
index ab1c0cd..67025cd 100644
--- a/setup.py
+++ b/setup.py
@@ -62,6 +62,12 @@ setup(
         'Programming Language :: Python :: 3.3',
         'Programming Language :: Python :: 3.4',
         'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: 3.7',
+        'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
+        'Programming Language :: Python :: 3.11',
     ],
 
     # What does your project relate to?
-- 
2.35.1

